classes/migration/upgrade/v3_4_0/PreflightCheckMigration.php

```diff
@@ -31,745 +34,1135 @@ abstract protected function getContextTable(): string;
+                    )
+                )
+                ->whereNull('p.publication_id')
+                ->pluck('s.submission_id', 'last.publication_id');
+            foreach ($rows as $publicationId => $submissionId) {
+                $this->_installer->log("The current publication ID ({$publicationId}) for the submission ID {$submissionId} is invalid, the publication ID {$publicationId} will replace it");
+                $affectedRows += DB::table('submissions')->where('submission_id', '=', $submissionId)->update(['current_publication_id' => $publicationId]);
             }
 
+            // The current_publication_id is nullable, but it's in fact required by the software, so we delete orphan entries instead of nulling them
+            $affectedRows += $this->deleteRequiredReference('submissions', 'current_publication_id', 'publications', 'publication_id');
+            return $affectedRows;
+        });
+
+        $this->addTableProcessor('submission_files', function (): int {
+            $affectedRows = 0;
+            // Depends directly on ~5 entities: file_id->files.file_id genre_id->genres.genre_id source_submission_file_id->submission_files.submission_file_id submission_id->submissions.submission_id uploader_user_id->users.user_id
+            $affectedRows += $this->deleteRequiredReference('submission_files', 'submission_id', 'submissions', 'submission_id');
+            $affectedRows += $this->deleteRequiredReference('submission_files', 'file_id', 'files', 'file_id');
+            $affectedRows += $this->cleanOptionalReference('submission_files', 'uploader_user_id', 'users', 'user_id');
+            $affectedRows += $this->cleanOptionalReference('submission_files', 'source_submission_file_id', 'submission_files', 'submission_file_id');
+            $affectedRows += $this->cleanOptionalReference('submission_files', 'genre_id', 'genres', 'genre_id');
+            return $affectedRows;
+        });
+
+        $this->addTableProcessor('publications', function (): int {
+            $affectedRows = 0;
+            // Depends directly on ~4 entities: primary_contact_id->authors.author_id doi_id->dois.doi_id(not found in previous version) section_id->sections.section_id submission_id->submissions.submission_id
+            $affectedRows += $this->deleteRequiredReference('publications', 'submission_id', 'submissions', 'submission_id');
+            $affectedRows += $this->cleanOptionalReference('publications', 'primary_contact_id', 'authors', 'author_id');
+            return $affectedRows;
+        });
+
+        $this->addTableProcessor('user_groups', function (): int {
+            $affectedRows = 0;
+            // Depends directly on ~1 entities: context_id->context_table.context_id
+            // Custom field (not found in at least one of the softwares)
+            $affectedRows += $this->cleanOptionalReference('user_groups', 'context_id', $this->getContextTable(), $this->getContextKeyField(), $this->ignoreZero('context_id'));
```

Comment:
I think we [still have this problem](https://github.com/pkp/pkp-lib/pull/8970#discussion_r1207505674) with the preflight check trying to update `context_id` to `null`. I get:
```
ERROR: Upgrade failed: DB: SQLSTATE[23000]: Integrity constraint violation: 1048 Column 'context_id' cannot be null (SQL: update `user_groups` set `context_id` = ? where `context_id` in (180))
```
This is a case where an entry in `journals` was deleted (`journal_id = 180`) and data in `user_groups` was not cleaned up.

Inspecting the code, it looks like there are a couple of problems:
- The `cleanOptionalReference` function expects to set missing references to `null`, but in this case that will generate bad data. `context_id` should only be `null` in `user_groups` for the administrator role, but this will result in numerous other entries getting set to `null` where their respective `journal_id` entries are missing.
- The `context_id` column will not be nullable anyway at this point in the upgrade, I don't think.

Reply:
That's right! I thought that I kept all changes related to the extra FKs on two separated commits, and overlooked the rebased modifications. I'll regenerate the automated code using a fresh database and do a diff, it should be enough to capture the divergences.

Reply:
I've updated the PR. After regenerating the cleanup code, I flagged one more divergence related to recent changes on the nullability of the `event_log.user_id`